API.pm

```diff
@@ -759,15 +757,22 @@ sub _pagingGet {
 
 	my $limit = $params->{limit};
 	$params->{limit} = min($params->{limit}, QOBUZ_LIMIT);
+	
+	my $extractor = ref $params->{_extractor} ? '' : $params->{_extractor};
 
-	my $getMaxFn = ref $params->{_maxKey} ? delete $params->{_maxKey} : sub {
-		my ($results) = @_;
-		$results->{$params->{_maxKey}}->{total};
+	my $getMaxFn = ref $params->{_maxKey} ? $params->{_maxKey} : sub {
+		my $results = shift;
+		$results->{$params->{_maxKey} || $extractor}->{total};
+	};
```

Comment:
Instead of re-using the `_extractor` to get the max, and removing the `_maxKey`, and then modifying the `pagingGet()` to accommodate for these different uses of the same thing, wouldn't it make sense and be much easier to just fix `_maxKey` for the playlists?

Reply:
Komisch, plötzlich war mein Kommentar weg obwohl mehrfache editiert. Stehe mit GitHub noch auf Kriegsfuß.

Hier nochmal:

Nein, der Grund ist ein Anderer, Du musst eine Zeile früher einsteigen, dann wird das klar. 

my $extractor = ref $params->{_extractor} ? '' : $params->{_extractor};

Im Falle dass _extractor ein Funktionspointer ist, wird $extractor = '' ansonst enthält er ja einen String mit dem Namen des Wertes den ich auslesen will.

Das hier ist nur eine Kodevereinfachung für die Aufrufer von _pagingGet. Es reicht dann völlig einmal für _extractor = 'tracks' zu setzen und man muss das nicht nochmal wir _maxKey und und _limitKey wiederholen. Ich habe mich willkürlich für _extractor entschieden, weil der im bisherigen Aufruferkode immer als erstes kam.

Ich benutze in meiner Plugin Version überhaupt keine Funktionspointer mehr für _pagingGet() und in allen meinen Aufrufen müsste ich dann den Valuenamen immer dreimal zuweisen, was irgendwie blöd aussieht. 

Wenn Du dir getPlaylistTracks() anschaust, siehst Du das ich _maxKey ganz rausgenommen habe, dass kann nur funktionieren wenn _extractor ein Variablenname enthält und _pagingGet() weiß wenn _maxKey fehlt er den gleichen Name wie _extractor nehmen soll. Das ist der Sinn des Ganzen.

Die ganze Konstruktion mit der Möglichkeit eigene Funktionen festzulegen ist doch eigentlich overkill.
Gebrauch wird das aktuell nur in  getUserFavorites() da ihr da alle Favoriten auf einmal einlest um sie später dann wieder zu zerlegen nach Alben, Tracks und Interpreten. Das mache ich anders, ich lese nur das ein, was der Benutzer sehen will und die API kennt ja den Parameter 'type' wo ich festlegen kann welche Favoriten ich abfrage.
Eigentlich müsste man nur den Namen festlegen wo die Items drin stehen denn die Namen der Variablen fürs Paging sind in der API doch festgelegt. (limit, total, offset, items)

Reply:
Du magst es für Overkill halten, aber die Mehrfachnutzung eines Wertes für unterschiedliche Dinge ist unschön 😁. Manchmal ist es einfach sauberer, getrennte Methoden zu haben, weil es sonst sicherlich früher oder später zu ungewollten Nebeneffekten kommt.

Reply:
>>>
Manchmal ist es einfach sauberer, getrennte Methoden zu haben, weil es sonst sicherlich früher oder später zu ungewollten Nebeneffekten kommt.
<<<

Dem würde ich doch nie widersprechen, dass sehe ich ganz genauso.

In Falle von _pagingGet() habe ich das erst einmal nur wegen der Kompatibilität der Routine gemacht. Also falls jemand _pagingGet() in seinem Code aufruft soll sich die Funktion genauso verhalten wie vorher.  Aber eigentlich müssten wir darauf keine Rücksicht nehmen, denn wer benutzt außer uns schon _paginGet() :-).
Und ja, meine Lösung war vielleicht pragmatisch, aber schon etwas schmutzig in der Umsetzung. Aber ist die Verwendung unterschiedlicher Datentypen für ein und den gleichen Parameter den sauberer?
Sauberer wäre es sicherlich einen neuen Hashwert z.B. '_typeKey' einzuführen, der immer nur den Namen des Ergebnishashes enthält den man abfragt.  Der ist nämlich immer gleich, egal ob ich 'items, limit, total oder offset danach abfrage. _extractor ist nicht anderes als eine Abfrage der items, _maxKey eben von total. Das ist so, wenn Du willst schicke ich dir Dumps die das zeigen. 

Mein Einwand mit der etwas flapsigen Bezeichnung 'overkill' bezog sich aber nicht auf das Thema der Mehrfachnutzung, sondern darauf, dass _pagingGet() mehr Funktionaliät bereitstellt, als wir für die gewünschten Ergebnisse benötigen.

Ich könnte dir praktisch zeigen, dass man vollständig ohne Funktionspointer als Parameter für _paginGet() auskommt . _pagingGet() ruft ja intern _get() auf und benötigt nur einen Stringparameter um alles zu steuern und das Plugin macht was es soll. Dadurch wird der Kode von _pagingGet() natürlich einfacher.

Das heißt nicht, dass der jetzige Kode falsch ist, er ist nur umfangreicher als er sein müsste.

Dann müsste allerdings die Funktion getUserFavorites() nebst ihren Aufrufern bei Dir angepasst werden. Deshalb habe ich es im Fix erstmal so gelassen damit es nicht zu viele Änderungen auf einmal sind.
Du siehst ja selbst was es hier jetzt schon für einen Aufwand macht, das war eigentlich nicht mein Gedanke.
Vielleicht ist es besser und für alle einfacher Du änderst es nach Deinen Vorstellungen ab und gut ist es.
Das ursächliche Problem, dass nur 100 Einträge kommen ist ja gelöst und meine Plugin Version zeigt ja, dass es funktioniert.